### ISSUE TITLE
Assertion failure in `shared_heap.c` in linuxO0 CI run

### ISSUE THREAD

**Issue by kayceesrk** [ocaml/ocaml#13090]
There is an unrelated assertion failure in #13079 in the linux-O0 CI run while testing the bytecode version of `mctest.ml`. See

https://github.com/ocaml/ocaml/actions/runs/8623935626/job/23638056562?pr=13079#step:7:77

```
> [00] file runtime/shared_heap.c; line 1398 ### Assertion failed: local->stats.pool_live_words == pool_stats.live
```

Copying the relevant snippet here as the CI logs will be culled eventually:
```
Process 12822 got signal 6(Aborted), core dumped
Could not find core file.
 ... testing 'mctest.ml' => failed
 ... testing 'mctest.ml' with line 3 (hasunix) => passed
 ... testing 'mctest.ml' with line 5 (bytecode) => failed (Running program /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte without any argument: command
/home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte 
failed with exit code -6)
 ... testing 'mctest.ml' with line 7 (native) => passed
> Specified modules: mctest.ml
> Source modules: mctest.ml
> Running test hasunix with 1 actions
> 
> Running action 1/1 (hasunix)
> Action 1/1 (hasunix) => passed (unix library available)
> Running test bytecode with 9 actions
> 
> Running action 1/9 (setup-ocamlc.byte-build-env)
> Action 1/9 (setup-ocamlc.byte-build-env) => passed
> 
> Running action 2/9 (ocamlc.byte)
> Compiling program /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte from modules  mctest.ml
> Commandline: /home/runner/work/ocaml/ocaml/runtime/ocamlrund /home/runner/work/ocaml/ocaml/ocamlc -use-runtime /home/runner/work/ocaml/ocaml/runtime/ocamlrund -I /home/runner/work/ocaml/ocaml/runtime -nostdlib -I /home/runner/work/ocaml/ocaml/stdlib -I /home/runner/work/ocaml/ocaml/otherlibs/unix unix.cma -o /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte mctest.ml
>   Redirecting stdout to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/ocamlc.byte.output 
>   Redirecting stderr to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/ocamlc.byte.output 
> Action 2/9 (ocamlc.byte) => passed
> 
> Running action 3/9 (check-ocamlc.byte-output)
> Comparing compiler output /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/ocamlc.byte.output to reference /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/mctest.compilers.reference
> Action 3/9 (check-ocamlc.byte-output) => passed
> 
> Running action 4/9 (run)
> Commandline: /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte
>   Redirecting stdout to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte.output 
>   Redirecting stderr to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte.output 
> ### begin stdout ###
> [00] file runtime/shared_heap.c; line 1398 ### Assertion failed: local->stats.pool_live_words == pool_stats.live
> ### end stdout ###
> Action 4/9 (run) => failed (Running program /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte without any argument: command
> /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlc.byte/mctest.byte 
> failed with exit code -6)
> Running test native with 8 actions
> 
> Running action 1/8 (setup-ocamlopt.byte-build-env)
> Action 1/8 (setup-ocamlopt.byte-build-env) => passed
> 
> Running action 2/8 (ocamlopt.byte)
> Compiling program /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/mctest.opt from modules  mctest.ml
> Commandline: /home/runner/work/ocaml/ocaml/runtime/ocamlrund /home/runner/work/ocaml/ocaml/ocamlopt -runtime-variant d -I /home/runner/work/ocaml/ocaml/runtime -nostdlib -I /home/runner/work/ocaml/ocaml/stdlib -I /home/runner/work/ocaml/ocaml/otherlibs/unix unix.cmxa -o /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/mctest.opt mctest.ml
>   Redirecting stdout to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/ocamlopt.byte.output 
>   Redirecting stderr to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/ocamlopt.byte.output 
> Action 2/8 (ocamlopt.byte) => passed
> 
> Running action 3/8 (check-ocamlopt.byte-output)
> Comparing compiler output /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/ocamlopt.byte.output to reference /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/mctest.compilers.reference
> Action 3/8 (check-ocamlopt.byte-output) => passed
> 
> Running action 4/8 (run)
> Commandline: /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/mctest.opt
>   Redirecting stdout to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/mctest.opt.output 
>   Redirecting stderr to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/mctest.opt.output 
> ### begin stdout ###
> done [100](https://github.com/ocaml/ocaml/actions/runs/8623935626/job/23638056562?pr=13079#step:7:101)000000
> done 100000000
> done 100000000
> done 100000000
> ### end stdout ###
> Action 4/8 (run) => passed
> 
> Running action 5/8 (check-program-output)
> Comparing program output /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.byte/mctest.opt.output to reference /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/mctest.reference
> Action 5/8 (check-program-output) => passed
> 
> Running action 6/8 (setup-ocamlopt.opt-build-env)
> Action 6/8 (setup-ocamlopt.opt-build-env) => passed
> 
> Running action 7/8 (ocamlopt.opt)
> Compiling program /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.opt/mctest.opt from modules  mctest.ml
> Commandline: /home/runner/work/ocaml/ocaml/ocamlopt.opt -runtime-variant d -I /home/runner/work/ocaml/ocaml/runtime -nostdlib -I /home/runner/work/ocaml/ocaml/stdlib -I /home/runner/work/ocaml/ocaml/otherlibs/unix unix.cmxa -o /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.opt/mctest.opt mctest.ml
>   Redirecting stdout to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.opt/ocamlopt.opt.output 
>   Redirecting stderr to /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.opt/ocamlopt.opt.output 
> Action 7/8 (ocamlopt.opt) => passed
> 
> Running action 8/8 (check-ocamlopt.opt-output)
> Comparing compiler output /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/_ocamltest/tests/parallel/mctest/ocamlopt.opt/ocamlopt.opt.output to reference /home/runner/work/ocaml/ocaml/testsuite/tests/parallel/mctest.compilers.reference
> Action 8/8 (check-ocamlopt.opt-output) => passed
```

**Comment by gasche** [comment]
Test source: https://github.com/ocaml/ocaml/blob/4c6a3849022ba19c23fb1860095f65eb09da157c/testsuite/tests/parallel/mctest.ml

I wondered if this could be related to an orphaning issue. Interestingly, this benchmark spawns several domains, but it never joins any of them. (We had a discussion recently on whether that was an okay thing to do.) It looks like the worker domains never terminate either.

**Comment by kayceesrk** [comment]
Unsure of what is the source of this failure. I've pinged the usual suspects at Tarides, and none of them have seen this assertion failure previously. The first step would be to recreate this failure.

**Comment by jmid** [comment]
Just spotted this before opening a new issue for https://github.com/ocaml-multicore/multicoretests/issues/470

With the help of shrinking, I've been able to create the following reproducer:
```ocaml
let array_length = 2

type t = { mutable lists : char list array; }

let test () =
  let t = { lists = Array.make array_length [] } in
  t.lists.(0) <- List.init 256 (fun _ -> 'a');
  t.lists.(1) <- List.init 1024 (fun _ -> 'a');

  let wait = Atomic.make true in
  let d1 = Domain.spawn
      (fun () ->
         while Atomic.get wait do Domain.cpu_relax() done;
         ()) in
  let d2 = Domain.spawn
      (fun () -> Atomic.set wait false;
        t.lists.(0) <- List.rev t.lists.(0);
        Gc.compact ();
        t.lists.(1) <- List.rev t.lists.(1);
      ) in
  let () = Domain.join d1 in
  let () = Domain.join d2 in
  t.lists <- [| |];
  ()

let _ =
  for i = 1 to 10_000 do
    (if i mod 1000 = 0 then Printf.printf ".%!");
    test ()
  done
```

On my Linux box this reliably reproduces with both 5.2.0 and `trunk` from today:
```
/tmp$ ocamlopt -g -runtime-variant=d gcrepro.ml 
/tmp$ OCAMLRUNPARAM="s=2048,v=0,V=1" ./a.out 
[00] file runtime/shared_heap.c; line 1392 ### Assertion failed: local->stats.pool_live_words == pool_stats.live
Illegal instruction (core dumped)
```

Note: it seems easiest to trigger with the use of `Gc.compact`, but as the `testsuite/tests/parallel/mctest.ml` above indicates, compaction is probably not to blame. I can also see if I can cook up a reliable reproducer not involving compaction.

**Comment by jmid** [comment]
I've looked a bit more into this.
By instead spawning 3 domains in each iteration, with 2/3 doing nothing but spinning, I get something that fails almost immediately locally on 5.2.0, 5.2.1, 5.3.0, and trunk from today, i.e., since `Gc.compact` was restored.

Here's the revised reproducer:
```ocaml
type t = { mutable lists : char list array; }

let test () =
  let t = { lists = Array.make 2 [] } in
  t.lists.(0) <- List.init 256 (fun _ -> 'a');
  t.lists.(1) <- List.init 1024 (fun _ -> 'a');

  let wait = Atomic.make true in
  let d0 = Domain.spawn
      (fun () ->
         while Atomic.get wait do Domain.cpu_relax() done;
         ()) in
  let d1 = Domain.spawn
      (fun () ->
         while Atomic.get wait do Domain.cpu_relax() done;
         ()) in
  let d2 = Domain.spawn
      (fun () -> Atomic.set wait false;
        t.lists.(0) <- List.rev t.lists.(0);
        Gc.compact ();
        t.lists.(1) <- List.rev t.lists.(1);
      ) in
  let () = Domain.join d0 in
  let () = Domain.join d1 in
  let () = Domain.join d2 in
  t.lists <- [| |];
  ()

let _ =
  for i = 1 to 100 do
    test ()
  done
```

With a little printf-debugging added
```diff
diff --git a/runtime/shared_heap.c b/runtime/shared_heap.c
index 8a23ecb2f0..f916b2b858 100644
--- a/runtime/shared_heap.c
+++ b/runtime/shared_heap.c
@@ -1454,6 +1454,9 @@ static void verify_swept (struct caml_heap_state* local) {
 
   /* Check stats are being computed correctly */
   CAMLassert(local->stats.pool_words == pool_stats.alloced);
+  if (local->stats.pool_live_words != pool_stats.live) {
+    printf("local->stats.pool_live_words is %i\npool_stats.live is %i\n", local->stats.pool_live_words, pool_stats.live);
+  }
   CAMLassert(local->stats.pool_live_words == pool_stats.live);
   CAMLassert(local->stats.pool_live_blocks == pool_stats.live_blocks);
   CAMLassert(local->stats.pool_frag_words == pool_stats.overhead);
```
this reveals example differences between the two fields:
```
$ ocamlopt -g -runtime-variant=d gcrepro.ml 
$ OCAMLRUNPARAM="s=2048,v=0,V=1" ./a.out 
local->stats.pool_live_words is 4173
pool_stats.live is 4164
[00] file runtime/shared_heap.c; line 1460 ### Assertion failed: local->stats.pool_live_words == pool_stats.live
Illegal instruction (core dumped)
$ OCAMLRUNPARAM="s=2048,v=0,V=1" ./a.out 
local->stats.pool_live_words is 2356
pool_stats.live is 2338
[00] file runtime/shared_heap.c; line 1460 ### Assertion failed: local->stats.pool_live_words == pool_stats.live
Illegal instruction (core dumped)
```

**Comment by jmid** [comment]
Despite spawning 3 domain (+ 3 backup threads), the backtraces I've seen so far have 2 domains + 2 backup threads running. Here's a representative example backtrace:
```
(gdb) thread apply all bt

Thread 50 (Thread 0x7fffef400640 (LWP 1507481) "a.out"):
#0  futex_wait (private=0, expected=2, futex_word=0x55555567fc70) at ../sysdeps/nptl/futex-internal.h:146
#1  __GI___lll_lock_wait (futex=futex@entry=0x55555567fc70, private=0) at ./nptl/lowlevellock.c:49
#2  0x00007ffff7c98002 in lll_mutex_lock_optimized (mutex=0x55555567fc70) at ./nptl/pthread_mutex_lock.c:48
#3  ___pthread_mutex_lock (mutex=mutex@entry=0x55555567fc70) at ./nptl/pthread_mutex_lock.c:93
#4  0x00005555555e31cf in caml_plat_lock_blocking (m=0x55555567fc70) at runtime/caml/platform.h:457
#5  backup_thread_func (v=0x55555567fbd0) at runtime/domain.c:1088
#6  0x00007ffff7c94ac3 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#7  0x00007ffff7d26850 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

Thread 49 (Thread 0x7ffff5a00640 (LWP 1507480) "a.out"):
#0  0x00005555556029ed in caml_gc_log (msg=msg@entry=0x55555561e2d0 "requesting stw empty_minor_heap") at runtime/misc.c:84
#1  0x000055555560258f in caml_try_empty_minor_heap_on_all_domains () at runtime/minor_gc.c:858
#2  0x0000555555602605 in caml_empty_minor_heaps_once () at runtime/minor_gc.c:882
#3  0x00005555555e401f in caml_domain_terminate (last=last@entry=false) at runtime/domain.c:2033
#4  0x00005555555e45a9 in domain_thread_func (v=<optimized out>) at runtime/domain.c:1267
#5  0x00007ffff7c94ac3 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#6  0x00007ffff7d26850 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

Thread 4 (Thread 0x7ffff6400640 (LWP 1507435) "a.out"):
#0  caml_failed_assert (expr=expr@entry=0x55555561edf0 "local->stats.pool_live_words == pool_stats.live", file_os=file_os@entry=0x55555561e98d "runtime/shared_heap.c", line=line@entry=1460) at runtime/misc.c:51
#1  0x0000555555607a4a in verify_swept (local=0x555555689740) at runtime/shared_heap.c:1460
#2  caml_sweep (local=0x555555689740, work=work@entry=512) at runtime/shared_heap.c:621
#3  0x00005555555f9c18 in major_collection_slice (howmuch=howmuch@entry=512, participant_count=participant_count@entry=0, barrier_participants=barrier_participants@entry=0x0, mode=mode@entry=Slice_opportunistic, force_compaction=force_compaction@entry=0) at runtime/major_gc.c:1796
#4  0x00005555555fa805 in caml_opportunistic_major_collection_slice (howmuch=howmuch@entry=512) at runtime/major_gc.c:1988
#5  0x0000555555601ffd in caml_do_opportunistic_major_slice (domain_state=0x555555688ba0, unused=0x0) at runtime/minor_gc.c:767
#6  caml_do_opportunistic_major_slice (unused=0x0, domain_state=0x555555688ba0) at runtime/minor_gc.c:758
#7  minor_gc_leave_barrier (participating_count=2, domain=0x555555688ba0) at runtime/minor_gc.c:739
#8  caml_empty_minor_heap_promote (domain=domain@entry=0x555555688ba0, participating_count=participating_count@entry=2, participating=participating@entry=0x5555556881c0) at runtime/minor_gc.c:697
#9  0x00005555556022c8 in caml_stw_empty_minor_heap_no_major_slice (domain=0x555555688ba0, participating_count=2, participating=0x5555556881c0, unused=<optimized out>) at runtime/minor_gc.c:799
#10 0x00005555555e3074 in stw_handler (domain=0x555555688ba0) at runtime/domain.c:1507
#11 handle_incoming (s=0x55555567f9c8) at runtime/domain.c:360
#12 0x00005555555e32e9 in caml_handle_incoming_interrupts () at runtime/domain.c:373
#13 backup_thread_func (v=0x55555567f9b0) at runtime/domain.c:1069
#14 0x00007ffff7c94ac3 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#15 0x00007ffff7d26850 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

Thread 1 (Thread 0x7ffff7eba740 (LWP 1507430) "a.out"):
#0  futex_wait (private=0, expected=2, futex_word=0x5555556d3a00) at ../sysdeps/nptl/futex-internal.h:146
#1  __GI___lll_lock_wait (futex=futex@entry=0x5555556d3a00, private=0) at ./nptl/lowlevellock.c:49
#2  0x00007ffff7c972a3 in __pthread_mutex_cond_lock (mutex=mutex@entry=0x5555556d3a00) at ../nptl/pthread_mutex_lock.c:93
#3  0x00007ffff7c93934 in __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x5555556d3a00, cond=0x5555556d3b10) at ./nptl/pthread_cond_wait.c:616
#4  ___pthread_cond_wait (cond=cond@entry=0x5555556d3b10, mutex=mutex@entry=0x5555556d3a00) at ./nptl/pthread_cond_wait.c:627
#5  0x000055555560b8f3 in sync_condvar_wait (m=0x5555556d3a00, c=0x5555556d3b10) at runtime/sync_posix.h:116
#6  caml_ml_condition_wait (wcond=<optimized out>, wmut=<optimized out>) at runtime/sync.c:193
#7  <signal handler called>
#8  0x00005555555c283e in camlStdlib__Domain$loop_752 () at domain.ml:292
#9  0x00005555555c194e in camlStdlib__Mutex$protect_277 () at mutex.ml:28
#10 0x00005555555c27dc in camlStdlib__Domain$join_749 () at domain.ml:297
#11 0x00005555555a12f1 in camlGcrepro$test_273 () at gcrepro.ml:25
#12 0x00005555555a15af in camlGcrepro$entry () at gcrepro.ml:31
#13 0x000055555559ebec in caml_program ()
#14 <signal handler called>
#15 0x00005555556112cc in caml_startup_common (pooling=<optimized out>, argv=<optimized out>) at runtime/startup_nat.c:127
#16 caml_startup_common (argv=<optimized out>, pooling=<optimized out>) at runtime/startup_nat.c:86
#17 0x000055555561135f in caml_startup_exn (argv=<optimized out>) at runtime/startup_nat.c:134
#18 caml_startup (argv=<optimized out>) at runtime/startup_nat.c:139
#19 caml_main (argv=<optimized out>) at runtime/startup_nat.c:146
#20 0x000055555559e952 in main (argc=<optimized out>, argv=<optimized out>) at runtime/main.c:37
```

To me, this looks like
- a waiting backup thread
- a child domain terminating
- a backup thread serving an incoming request - and triggering the assertion failure
- a main domain waiting on a `join`

**Comment by gasche** [comment]
Currently the allocation and deallocation logic (incrementally) maintain statistics in `local->stats`. In debug mode, in `verify_*` functions, we periodically recompute statistics from scratch by traversing all memory pools, and we check that the two values match. This is the assertion that fails.

My impression is that the compactor does not update pool statistics when it evacuates blocks from one pool to another, so the pool stats become incorrect at this point.

I would do the update around here (increment the stats of the destination pool):

https://github.com/ocaml/ocaml/blob/2da20f7641342c880289c850e83d4f2f58ba23c8/runtime/shared_heap.c#L1233-L1234

I'm not familiar with the compaction code, but I don't think we should try to decrement the stats for the pool that is the evacuation source. Two reasons:

1. The block is not immediately made free, it stays around as a forwarding pointer after evacuation (which would count as non-free if we did verify the statistics at this point, which I don't think we do).
2. The evacuated pool are freed towards the end of compaction, so their blocks do not count in the statistics anymore by the end of compaction.

**Comment by sadiqj** [comment]
I have a fix for this in https://github.com/sadiqj/ocaml/commit/dfafaf77296737cc554ba3e29e065475c82a3066 . With that commit, @jmid 's reproduction no longer fails on the debug runtime with heap verification.

We essentially didn't account for the stats when we encountered garbage in a pool being evacuated from. This wasn't caught because it's very difficult to actually end up with garbage in the major heap after three major cycles.

What I'm currently trying to figure out is _why_ we have garbage in the major heap. Backwards-stepping through some failures that logged out garbage block locations with rr, I've not seen anything being allocated in finalisers.

Will spend some more time digging.

**Comment by jmid** [comment]
> I have a fix for this in [sadiqj@dfafaf7](https://github.com/sadiqj/ocaml/commit/dfafaf77296737cc554ba3e29e065475c82a3066) . With that commit, [@jmid](https://github.com/jmid) 's reproduction no longer fails on the debug runtime with heap verification.

Cool. I just did a local build from that branch and I can confirm that the larger Gc test from which the reproducer came is no longer able to trigger the assertion failure 🙂👍

**Comment by gasche** [comment]
[Earlier](https://github.com/ocaml/ocaml/issues/13090#issuecomment-2596102454) I made a hypothesis about what was going wrong, which is different from your current explanation: I guessed that the issue was that compaction moves blocks from one pool to another without updating the stats. The fact that your orthogonal fix removes the issue means that my guess was wrong, but it took me some time to realize why.

My understanding now is that the compactor only moves block across pools that are owned by the same domain, never across domains. Because the statistics we maintain are per-domain (per-heap) and not per-pool, there is no need to update statistics when migrating blocks across same-domain pools. On the other hand, in the conditional branch where a block is collected instead of migrated, then indeed we need to update the stats -- as your patch in #13774 is doing.

**Comment by Octachron** [comment]
There was another instance of this assertion failure in `mctest.ml` yesterday on a CI run in #13302 .
